eth-bytecode-db/verifier-alliance-database/src/helpers.rs

```diff
@@ -0,0 +1,63 @@
+macro_rules! insert_then_select {
+    ( $txn:expr, $entity_module:ident, $active_model:expr, $update_on_conflict:expr, [ $( ($column:ident, $value:expr) ),+ $(,)? ] ) => {
+        {
+            use anyhow::Context;
+            use sea_orm::{ColumnTrait, EntityTrait, QueryFilter};
+
+            let result: Result<_, sea_orm::DbErr> = $entity_module::Entity::insert($active_model.clone())
+                .on_conflict(sea_orm::sea_query::OnConflict::new().do_nothing().to_owned())
+                .exec($txn)
+                .await;
+
+            // Returns the model and the bool flag showing whether the model was actually inserted.
+            match result {
+                Ok(res) => {
+                    let last_insert_id = res.last_insert_id;
+                    let model = $entity_module::Entity::find_by_id(last_insert_id.clone())
+                        .one($txn)
+                        .await
+                        .context(format!("select from \"{}\" by \"id\"", stringify!($entity_module)))?
+                        .ok_or(anyhow::anyhow!(
+                            "select from \"{}\" by \"id\"={:?} returned no data",
+                            stringify!($entity_module),
+                            last_insert_id
+                        ))?;
+
+                    Ok((model, true))
+                }
+                Err(sea_orm::DbErr::RecordNotInserted) => {
+                    let mut model =
+                        $entity_module::Entity::find()
+                            $(
+                                .filter($entity_module::Column::$column.eq($value))
+                            )*
+                            .one($txn)
+                            .await
+                            .context(format!("select from \"{}\" by unique columns", stringify!($entity_module)))?
+                            .ok_or(anyhow::anyhow!("select from \"{}\" by unique columns returned no data", stringify!($entity_module)))?;
+                    // The active model have not been inserted.
+                    // Thus, there were a value already that we need to update.
+                    if $update_on_conflict {
```

Comment:
Can you explain why is this custom logic needed? I seem to miss the point 😅

Reply:
This macro definition was just copied from `eth-bytecode-db` crate. Inside `eth-bytecode-db` there were both cases - when inserted data should override existing ones and when shouldn't. I don't think we update anything inside `verifier-alliance-database` crate, but I'm planning to remove the copypaste and just reuse one macro definition from both crates later, so decided not to update the interface here

Reply:
As far as I remember it requires the columns to be specified. But in our case some of uniqueness of some of the columns was specified by their hash instead of themselves. That resulted in `.on_conflict` option being not working. I will check if that is still the case in the next PRs